### PR TITLE
Implement recent commits graph

### PR DESCRIPTION
**Pull request by sahinakkaya**
### Overview
This is the implementation of Recent Commits page. This feature was mentioned on #18262.

It adds another tab to Activity page called Recent Commits. Recent Commits tab shows number of commits since last year for the repository.

Before:
image

After:
image

---

#### Features
- See number of commits for each week for the past year
- Hover over the graph to see a detailed tooltip for that week
- Use the same cache from Contributors page so that the loading of data will be fast once it is cached by visiting either one of the pages

#### Issues to consider before merging:
- Since only last year's data is necessary, we don't need to process whole history of the repository. This implementation currently uses data from "Contributors" page which processes whole history of the repo. If the data is cached before, we are good. If it is not cached, we wait a little bit longer to process whole history. This issue is not important if user wants to visit other pages (contributors, [code frequency](https://github.com/go-gitea/gitea/pull/29191)) because they will need whole data any way.

### CODE REVIEW

File: web_src/js/components/RepoRecentCommits.vue
Comment (sahinakkaya) [inline web_src/js/components/RepoRecentCommits.vue]: These lines are duplicated. In [Code Frequency graph PR](https://github.com/go-gitea/gitea/pull/29191), we extracted these to a different file to remove duplicate code. Once that PR is merged, I will make use of it to remove these lines.

File: web_src/js/components/RepoRecentCommits.vue
Comment (sahinakkaya) [inline web_src/js/components/RepoRecentCommits.vue]: And this one too.

File: options/locale/locale_en-US.ini
Comment (denyskon) [inline options/locale/locale_en-US.ini]: Please never ever duplicate translations that already exist! You can rename the ones made for commit graph and use them here. Also, use placeholders if the only thing that changes is the type of data being loaded ("Could not load %s")

**Review by denyskon — CHANGES_REQUESTED** [review]

File: web_src/js/components/RepoRecentCommits.vue
Comment (sahinakkaya) [inline web_src/js/components/RepoRecentCommits.vue:126]: I am using these translations here. How am I supposed to use translations with "%s" inside vue template @denyskon ?

File: web_src/js/features/recent-commits.js
Comment (sahinakkaya) [inline web_src/js/features/recent-commits.js:19]: Here is where they are coming from. Can we pass arguments here? @denyskon

Comment (sahinakkaya) [reply web_src/js/features/recent-commits.js:19]: Oh, wait. I think I got it.

Comment (sahinakkaya) [reply web_src/js/features/recent-commits.js:19]: Nope, I didn't. See https://github.com/go-gitea/gitea/pull/29210#discussion_r1493414014

File: templates/repo/recent_commits.tmpl
Comment (sahinakkaya) [inline templates/repo/recent_commits.tmpl]: How do I pass arguments for translations with placeholders here?

Comment (denyskon) [inline templates/repo/recent_commits.tmpl]: Simply put it after the name of the locale key, separated with a whitespace. You can also use another locale string:

```suggestion
		data-locale-component-failed-to-load="{{ctx.Locale.Tr "graphs.component_failed_to_load" (ctx.Locale.Tr "repo.recent_commits")}}"
```

Comment (denyskon) [reply templates/repo/recent_commits.tmpl]: As it looks to me like we could use such messages in various places, I'd create a separate group for that.

```ini
[graphs]
component_failed_to_load="An unexpected error happened when loading %s"
```

**Comment by sahinakkaya** [comment]
@denyskon can you remove lgtm/blocked tag if it is ok now?

**Comment by denyskon** [comment]
Sadly the only way to do it for me would be to approve your PR, and currently I don't have time to test it properly. Maintainers with merge permissions are able to dismiss my review.....

**Comment by delvh** [comment]
Surprisingly we can't either.
Sounds like a GitHub bug to me

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Merge https://github.com/go-gitea/gitea/pull/29191 before this and then alter this one to use the new shared functions.

**Review by denyskon — APPROVED** [review]
(assuming the deduplication after merging the other PR)

File: web_src/js/components/RepoRecentCommits.vue
Comment (silverwind) [inline web_src/js/components/RepoRecentCommits.vue:148]: Seems too high imho, the same diagram on Github is only 130px. I would suggest something like 250px.

Comment (sahinakkaya) [reply web_src/js/components/RepoRecentCommits.vue:148]: Ok, I've changed it.

File: routers/web/repo/recent_commits.go
Comment (wxiaoguang) [inline routers/web/repo/recent_commits.go:25]:
```suggestion
```

I would suggest avoid using PageData here, just use `data-repo-link` to pass the data for the component (just like the locale strings), it makes it easier to maintain.

Comment (wxiaoguang) [reply routers/web/repo/recent_commits.go:25]: Or, at least, it should use a "namesapce" like the guideline says:  `ctx.PageData["myModuleData"] = map[]{...}`. https://docs.gitea.com/contributing/guidelines-frontend (hmm, the guideline is not clear enough, I think it could be improved later)

Otherwise, if there are more than one vue components in one page, putting all their data in PageData would cause conflicts.

Comment (sahinakkaya) [reply routers/web/repo/recent_commits.go:25]: > just use `data-repo-link` to pass the data for the component

How do I pass a dynamic value (repo link) inside a `tmpl` file? The locale strings are set inside `/templates/repo/recent_commits.tmpl`.

Comment (wxiaoguang) [reply routers/web/repo/recent_commits.go:25]: I think you have used it somewhere 😁  this is only a nit, doesn't block the LGTM/merge. I could propose some more fixes in following PRs.

image

Comment (sahinakkaya) [reply routers/web/repo/recent_commits.go:25]: Oh, wait. I can access `ctx` :D

**Review by wxiaoguang — APPROVED** [review]
Overall LGTM.

File: routers/web/repo/recent_commits.go
Comment (sahinakkaya) [reply routers/web/repo/recent_commits.go:25]: I've done [this](https://github.com/go-gitea/gitea/pull/29210/commits/f8771d5a9ee189bd6754f04e89329c2848a69101) but it says "`repoLink` is undefined". image

It seems that the value of repoLink is not set when I try to fetch data with 

 response = await GET(`${repoLink}/activity/recent-commits/data`)

 Any suggestions?

Comment (sahinakkaya) [reply routers/web/repo/recent_commits.go:25]: @wxiaoguang the issue is not fixed. Please see my comment above.

Comment (wxiaoguang) [reply routers/web/repo/recent_commits.go:25]: No worry, you could use the old approach and let's merge, I will take a look later.

**Pull request merged**
